### PR TITLE
fix: hotplug disk feature is not working with kubletPodsDir changed w…

### PR DESCRIPTION
**Pull request by nxf129**
…hen disk is a block device

**What this PR does / why we need it**:
fix: hotplug disk feature is not working with kubletPodsDir changed when disk is a block device

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #  #10202 

**Special notes for your reviewer**:

**Release note**:

```release-note
None
```

### CODE REVIEW

**Comment by nxf129** [comment]
> @nxf129, I think you will need to adapt the unit tests to the new changes.
ok, I will do it.

File: pkg/virt-handler/hotplug-disk/mount.go
Comment (vasiliy-ul) [inline pkg/virt-handler/hotplug-disk/mount.go:458]: Basically, you do not need to pass `kubeletPodsDir` as an argument here as it is already available as a member of `m *volumeMounter` struct:
```suggestion
func (m *volumeMounter) getSourceMajorMinor(sourceUID types.UID, volumeName string) (uint64, os.FileMode, error) {
	basePath, err := deviceBasePath(sourceUID, m.kubeletPodsDir)
```
In such way you will have a more compact diff and will not need so many changes in the unit tests.

File: pkg/virt-handler/hotplug-disk/mount.go
Comment (vasiliy-ul) [inline pkg/virt-handler/hotplug-disk/mount.go:49]: Quick grep shows that `sourcePodBasePath` is not actually used in the code. It is defined here and then redefined in tests:
```
$ grep -RnI sourcePodBasePath
pkg/virt-handler/hotplug-disk/mount.go:49:	sourcePodBasePath = func(podUID types.UID) (*safepath.Path, error) {
pkg/virt-handler/hotplug-disk/mount_test.go:72:	orgSourcePodBasePath   = sourcePodBasePath
pkg/virt-handler/hotplug-disk/mount_test.go:572:			sourcePodBasePath = orgSourcePodBasePath
pkg/virt-handler/hotplug-disk/mount_test.go:582:			sourcePodBasePath = func(podUID types.UID) (*safepath.Path, error) {
pkg/virt-handler/hotplug-disk/mount_test.go:603:			sourcePodBasePath = func(podUID types.UID) (*safepath.Path, error) {
pkg/virt-handler/hotplug-disk/mount_test.go:613:			sourcePodBasePath = func(podUID types.UID) (*safepath.Path, error) {
pkg/virt-handler/hotplug-disk/mount_test.go:630:			sourcePodBasePath = func(podUID types.UID) (*safepath.Path, error) {
pkg/virt-handler/hotplug-disk/mount_test.go:644:			sourcePodBasePath = func(podUID types.UID) (*safepath.Path, error) {
pkg/virt-handler/hotplug-disk/mount_test.go:660:			sourcePodBasePath = func(podUID types.UID) (*safepath.Path, error) {
pkg/virt-handler/hotplug-disk/mount_test.go:788:			sourcePodBasePath = orgSourcePodBasePath
pkg/virt-handler/hotplug-disk/mount_test.go:843:			sourcePodBasePath = func(podUID types.UID) (*safepath.Path, error) {
pkg/virt-handler/hotplug-disk/mount_test.go:973:			sourcePodBasePath = func(podUID types.UID) (*safepath.Path, error) {
```
I raised a PR to drop it https://github.com/kubevirt/kubevirt/pull/10216 (and also to clean up some other unused code). Probably, you do not need this change at all and can leave the function as it is. I think it will not affect your PR.

File: pkg/virt-handler/hotplug-disk/mount_test.go
Comment (vasiliy-ul) [inline pkg/virt-handler/hotplug-disk/mount_test.go:285]:
```suggestion
```

Since you do not create this directory explicitly and the tests do not complain, I would assume that in fact it is not used. In such case, could you instead pass an empty string `""` where needed? Just to make it clear.

**Review by vasiliy-ul — COMMENTED** [review]
/approve

The PR looks good from my perspective. If I remember correctly, in the past we had several similar tracked issues and even with people assigned to work on those. However, not sure if we had ready PRs (or am I wrong?). Anyway, leaving it for @awels and @alicefr to also take a look here.

**Comment by alicefr** [comment]
@vasiliy-ul I think the issues you mention were more related when kubelet pod directory is a symlink.

**Comment by alicefr** [comment]
@nxf129 is there any reason why to put the directory in NewVolumeMounter in pkg/virt-handler/hotplug-disk/mount.go and not using the NewHotplugDiskManager in [pkg/hotplug-disk/hotplug-disk.go](https://github.com/kubevirt/kubevirt/blob/main/pkg/hotplug-disk/hotplug-disk.go#L54)?

**Comment by alicefr** [comment]
> > @nxf129 is there any reason why to put the directory in NewVolumeMounter in pkg/virt-handler/hotplug-disk/mount.go and not using the NewHotplugDiskManager in [pkg/hotplug-disk/hotplug-disk.go](https://github.com/kubevirt/kubevirt/blob/main/pkg/hotplug-disk/hotplug-disk.go#L54)?
> 
> Thank you for your suggestion. My consideration is based on the following aspects
> 
> 1. I tried to use hotplugDiskManager. GetHotplugTargetPodPathOnHost (podUid) to get the pod in the path of the host, this operation will affect deviceBasePath uid parameter of the anonymous functions, As a result, use cases related to deviceBasePath in unit tests will fail
> 2. Modify NewVolumeMounter as a minimal change that will not have much impact on the original unit tests

Well, you could do a similar function as [GetHotplugTargetPodPathOnHost](https://github.com/kubevirt/kubevirt/blob/main/pkg/hotplug-disk/hotplug-disk.go#L71).

**Comment by lynn901** [comment]
this pr solve the same problem
https://github.com/kubevirt/kubevirt/pull/9778

**Comment by vasiliy-ul** [comment]
> Still need to make sure all the commits are signed, since the DCO bot is unhappy.

+1

@nxf129, there is a merge commit 52d23b6dc7e32ac12c93d7b31653fc0322a8159c without `Signed-off-by`. Could you maybe rebase your changes on top of the latest `main`? Thus, the merge commit will not be needed and this will make the DCO check happy.

Adding a hold to silence the bot. Feel free to unhold when this is fixed.

/hold

**Comment by vasiliy-ul** [comment]
@nxf129, seems now you have 2 merge commits... You need to run `git pull <upstream-remote> main --rebase` from your branch. Then fix the conflicts (if any). After that, the merge commits should disappear. Alternatively, you can squash all your changes into a single commit. It should also be fine, as the changes here look pretty atomic.

**Comment by vasiliy-ul** [comment]
@nxf129, thanks for squashing the commits! However, you still have some style issues in the code:

```diff
diff --git a/pkg/virt-handler/hotplug-disk/mount.go b/pkg/virt-handler/hotplug-disk/mount.go
index 4b6435382..971c8d55f 100644
--- a/pkg/virt-handler/hotplug-disk/mount.go
+++ b/pkg/virt-handler/hotplug-disk/mount.go
@@ -440,7 +440,7 @@ func (m *volumeMounter) mountBlockHotplugVolume(
 }
 
 func (m *volumeMounter) getSourceMajorMinor(sourceUID types.UID, volumeName string) (uint64, os.FileMode, error) {
-	basePath, err := deviceBasePath(sourceUID,m.kubeletPodsDir)
+	basePath, err := deviceBasePath(sourceUID, m.kubeletPodsDir)
 	if err != nil {
 		return 0, 0, err
 	}
```

Please fix those, otherwise the CI check will not pass. You can build your branch locally and the code style will be adjusted automatically as needed. You just need to add and commit the changed files.

**Review by vasiliy-ul — APPROVED** [review]
/unhold

All looks good now. Removing the hold. Thanks!

**Pull request merged**
